### DISCUSSION TITLE
Querydsl 날짜 Fomating 의 문제점으로 인한 변경사항

### DISCUSSION THREAD

**Discussion by Aleexender** [SCBJ-7/SCBJ-BE#133] [Announcements]
### 문제의 시발점
사실 몇일 전에 올렸어야하는걸 해당 기능이 끝나고 나서 이 글을 올린점 양해의 말씀 부탁드려용!

문제의 기원은 제가 2차 가격으로 변경해야하는 **시간 기능을 구현을 못한것이 시발점**이 되었었습니다!
그래서 처음에는 저혼자 최대한 날짜를 포맷팅해서 startDate 와 secondPeriod를 LocalDateTime 으로 변환 한후 2차 가격날짜가 나오게 되면 해당 날짜를 안지났으면 1차가격을 보여주고 지낫으면 2차가격을 보여줘야했었습니다! 

#### 정확한 문제점
그럼 어디가 잘못된 것인가??

정확한 에러는 신텍스 에러가 발생했습니다
물론 Date 에서 포맷팅을 할때 발생했다고 확신한것이
Date 포맷팅 로직을넣고나서 테스트를 실행하니 발생을 하였고
INTERVAL* <- 위와같이 별표시 때문에 발생하게 되었던 것이였습니다
INTERVAL 이 HIBERNATE 예약어여서 발생한거였어요!

#### 그럼 어떻게 해결하였느냐?
여기서 엄청 많은 시행착오와 고민이 있었습니다....
1. service 딴으로 뺴서 하면 가능하지 않겠나!? 
1.1  일단 수많은 데이터들을 검색하고 결과값을 가져오는 도메인으로 인식하여 100만건 을 기준으로 잡고 조회할시 비지니스로직으로 뺴게된다면 성능상 면에서 엄청난 손실을 예상하였기에 마지막딴으로 미룰려고 하였습니다
1.2 비지니스 로직으로 뺀다고 한들 Sorting 과정으로 인한 where 절을 어떻게 핸들링 할지 답을 몰랐었습니다.. 또한 sorting 하는 과정을 어떻게 핸들링 해야할지 아직 잘 모르겠더라구요 ㅠㅠ

2. Repository에서 tuple 로 해결

결과적으로는 2번을 사용하여서 문제를 해결하였습니다.... tuple로 진행하여 미리 값들을 가져온후 querydsl 의 tuple인터페이스를 이용하여 가져온 값들을 자바 언어로 가공하여 sorting 하였습니다.... 

#### 그럼 왜이렇게 많은 시간이 걸렸느냐?
1. querydsl 의 미숙함 - 처음으로 querydsl 로 사용을 하고 동적쿼리를 처음다루다 보니 수많은 시간이 걸린거같습니다.( 왜 요즘 다른 회사들이querydsl 을 사용안하는지 알거같더라구요....) 상당한 복잡한 필터링과 요구사항이 들어가다가 보니 코드수가 수없이 길어짐으로 저 아시는현직자 분이 이럴땐 querydsl 보단 네이티브 쿼리를 많이 사용한다고 하더라구요.... ㅠㅠ
2. 로직상 최선의 방법을 고민 - 이부분에서 정말로 많은 고민을 했던거 같습니다... 어떻게하면 최대한 빠르고 괜찮게 가져올수있을까 stackOverFlow 부터 querydsl 공식 문서 까지 전부다 읽어오고 하였었습니다... 또한 DataJpaTest에 대해서도 공부하며 많은 시행착오를 겪었다고 생각합니다.... 비지니스 로직으로도 빼보고 Data DTO 로도 만들어서도 해결해보고 DataJpaTest 를 Mysql 로도 만들어보고... 참 많은 시도들을 해봤지만 해당 로직이 정말로 최선의 방법이였던거같습니다. 이로 인해 다른 클래스를만들필요도 없이 그냥 Response로 따로 빼서 하면 그만이니까요....물론 어느정도의 비지니스 로직이 들어가있긴합니다.... ㅠㅠㅠ

여기서 질문이 생길거 같아서 미리 답변 드릴게요!
#### 그러면 서비스 로직이 Repository안에 들어가있는거 아닌가요?
맞습니다! 하지만 Tuple 타입을 사용하게 되는것은 querydsl 에서 제공하는 기능이고 여러개의 대상을 프로젝션할 때를 대비해서 만든 타입이기 때문에 가능한 repository안에서 핸들링해주는게 좋은 설계입니다!

https://velog.io/@mingsound21/Querydsl-16.-%ED%94%84%EB%A1%9C%EC%A0%9D%EC%85%98%EA%B3%BC-%EA%B2%B0%EA%B3%BC-%EB%B0%98%ED%99%98

기다림을 가져주시고 기다려주신 모든 팀원분들에게 감사드리며 정말로 3일동안 힘든 시간이였지만 이렇게 해결되고 나니정말로 기쁘네요! 한단계 성장한 기분입니다!

긴글 읽어주셔서 감사합니다!

**Comment by YurimYang** [comment]
querydsl 처음 사용해보셨음에도 불구하고 끝까지 책임감 있게 마무리해주셔서 정말 감사합니다 👍  
옆에서 로직상 고민 많이 하시는 거 보면서 저도 제 코드에 대해 다시 한번 생각해보는 기회를 가질 수 있었어요 !! 너무 수고하셨습니다 👍
